mlir/lib/Gradient/Transforms/BufferizableOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,622 @@
+
+        // Construct the bufferized function type.
+        SmallVector<Type> argTypes;
+        for (const auto &it : llvm::enumerate(funcType.getInputs())) {
+            Type argType = it.value();
+            if (dyn_cast<TensorType>(argType)) {
+                argTypes.push_back(getBufferizedFunctionArgType(forwardOp, it.index(), options));
+                continue;
+            }
+            argTypes.push_back(argType);
+        }
+
+        ReturnOp returnOp = getAssumedUniqueReturnOp(forwardOp);
+        assert(returnOp && "expected func with single return op");
+        Location loc = returnOp.getLoc();
+
+        // 1. Bufferize every block.
+        for (Block &block : forwardOp.getBody()) {
+            if (failed(bufferization::bufferizeBlockSignature(&block, rewriter, options))) {
+                return failure();
+            }
+        }
+
+        // 2. For each result, keep track of which inplace argument it reuses.
+        SmallVector<Value> returnValues;
+        for (OpOperand &returnOperand : returnOp->getOpOperands()) {
+            Value returnVal = returnOperand.get();
+            auto tensorType = dyn_cast<TensorType>(returnVal.getType());
+            rewriter.setInsertionPoint(returnOp);
+
+            // If not a tensor type just forward it.
+            if (!tensorType) {
+                returnValues.push_back(returnVal);
+                continue;
+            }
+
+            // Note: If `inferFunctionResultLayout = true`, cast are later folded
+            // away.
+            BaseMemRefType resultType = options.functionArgTypeConverterFn(
+                tensorType, *options.defaultMemorySpaceFn(tensorType),
+                dyn_cast<func::FuncOp>(*forwardOp), options);
+            Value toMemrefOp =
+                rewriter.create<bufferization::ToMemrefOp>(loc, resultType, returnVal);
+            returnValues.push_back(toMemrefOp);
+        }
+
+        // 3. Rewrite the terminator.
+        forwardOp.walk([&](ReturnOp returnOp) {
+            PatternRewriter::InsertionGuard guard(rewriter);
+            rewriter.setInsertionPoint(returnOp);
+            rewriter.replaceOpWithNewOp<ReturnOp>(returnOp, returnValues, returnOp.getEmpty());
+        });
+
+        // 4. Rewrite the FuncOp type to buffer form. Also preserve unused return types.
+        SmallVector<Type> returnTypes;
+        for (auto retTy : forwardOp.getResultTypes()) {
+            auto tensorType = dyn_cast<TensorType>(retTy);
+            BaseMemRefType resultType = options.functionArgTypeConverterFn(
+                tensorType, *options.defaultMemorySpaceFn(tensorType),
+                dyn_cast<func::FuncOp>(*forwardOp), options);
+            returnTypes.push_back(resultType);
+        }
+        forwardOp.setType(FunctionType::get(op->getContext(), argTypes, returnTypes));
+
+        return success();
+    }
+};
+
+// Bufferization of gradient.reverse.
+struct ReverseOpInterface
```

Comment:
And I guess everything above would also apply here once we resolve those comments.